### ISSUE TITLE
Ma

### ISSUE THREAD

**Issue by jdjakub** [d-cook/SomethingNew#15]

**Comment by d-cook** [comment]
I think this is partly why Pavel was suggesting that we look at / demonstrate other tools like MPS or LightTable.

**Comment by d-cook** [comment]
When we've talked about a "substrate" within this collaboration, we could be referring to one of two things.

The first is the literal mechanism or toolset or framework (etc.) that could be used to implement a self-bootstrapping software system. Examples include the HTML DOM, programming language constructs, or existing systems like Squeek or Lively Kernel.

The second is representing data (and elements of a self-hosted system) in a form that can inspected & manipulated in place (at runtime), and that makes the structure of the data (or elements) self-evident.

The first is an implementation-detail of the second, and does not "come with" if/when the system is bootstrapped from one underlying implementation to another. It is incidental (and ideally invisible) to the system that is implemented atop it.

The second is the actual structural essence of the system, which does not change if/when it is ported to another underlying implementation.

For example, the essence might be a system of objects in the form of key-value mappings, but the implementation of those objects might be in the form of HTML DOM elements or constructs in some programming language.

We should clarify which definition of "substrate" we mean. (I've almost always been referring to the essence / "second" definition).

**Comment by d-cook** [comment]
Building upon my previous comment, a meaningful set of goals may be to:
1. Determine a good / ideal substrate (essence) that would ensure a fully open-ended system to exist.
2. Determine which existing toolset(s) would be most ideal for implementing such a system *quickly and simply* (regardless of other merits or lack thereof). This also means whatever will best & most-quickly enhance the ability to (and experience / immersiveness of carrying out) the next step (3).
3. Re-implement the system in itself ("bootsrap" it). That is, the "essence substrate" also becomes the "implementation substrate" (toolset). The cool thing about this is that the very definition (mechanism) of what the essence-substrate is and how it works, is subject to its own modification (and via immersive experimentation)  -- which in large part is one of the ultimate goals of this thing.

I'll not speculate much further on the next steps after that, though it probably includes (re-)implementing the implementation layer at a low level.

**Comment by jdjakub** [comment]
Ah. I'm talking about the first type of "substrate"; the initial starting system. 

However, I am not sure about your second concept of "substrate" as an "essence". Because we want there to be no end to the development and growth of what we create, is it meaningful to talk of an "essence" as if it were something that would be fundamental and unchanging? I imagine that we will have a first "essence", then a second, then a third, and so on ... the "essence" will simply be "what the system currently is", no?

But obviously I'm glossing over what you're likely *really* getting at, which is the *starting* "essence" distinct from the "implementation detail" I've been wondering about. Assuming I understand you correctly, for me this is the OROM object system. But the system described in the paper is only meant to be minimally self-starting; it suffers from many flaws that I'd hope to fix by re-implementing the system in itself, version N using version N-1.

Is "the OROM object system" an example of an "essence" as you mean it?

**Comment by d-cook** [comment]
Yes, the "essence" I talked about is an *initial* bootstrappable-representation. One cannot know the best model until one is immersed in it, and able to explore & experiment in a direct manner (Bret Victor's [Media for Thinking the Unthinkable](http://worrydream.com/MediaForThinkingTheUnthinkable) also comes to mind). It's like how Alan Kay talks about designing "almost a new thing" as a prerequisite to even being able to think of / discover the "new thing".

That's also why we want to look at the quickest road to creating such an immersive experience. Exploration of [OROM](http://www.vpri.org/pdf/tr2006003a_objmod.pdf) is great because it's a fleshed-out model of an adaptable system which can potentially be made into a visually immersive interactive experience.

That being said, one thought I have on the direction I see things going is that if the structure/mechanism that allows an adaptable system to be adaptable, is itself adaptable, then you can evolve the system itself into a new thing (or alternate "almost new" things), rather than having a series of temporary experiments that evolve separate or that have to be thrown out to explore an alternate mechanism. OROM might be sufficient in this regard.

**Comment by jdjakub** [comment]
Yep, I see we're on the same page then. So in a nutshell the "hypothesis" I'm working on is "implementing OROM (initial essence) in the DOM using the JS console (initial implementation) is a good place to start".

The separation of essence from implementation is an productive initial division. However, I think this analysis needs properly taking further to bring out the real forces at work that could help us decide how feasible or "sufficient" this is.

There are various clues, but no systematic approach (yet). The "tools" metaphor is one of them. All the time, we speak of building X "in" or "with" Y. We take care to emphasise the importance of getting to the form of "X in / using / with X", i.e. the system implemented "in itself", as we see in the COLAs paper.

And these aren't unique to us; this bootstrapping technique is well-known in the "compilers" field -- deep in "language-centered" territory. They know about, say, implementing a compiler for the C language, *in* the C language. But -- while they might not consider it relevant to their practise -- **this is an incomplete picture of what's going on**.

One does not make a compiler with "the C language" alone. There is "the rest" -- the "context" -- may I call it the "environment"? At first glance, I want to characterise it as simply the **runtime** environment plus the **compile-time** environment. The first is the up-and-running system being made, the second is the system acting as the "fabrication tool".

(ahem -- and it seems that the runtime half might not be that relevant to what happens *before the runtime being made even exists* ... i.e. only the *compile-time* environment is relevant to language-bootstrapping, which consists of at least a few distinguished sub-components, which need bringing into the spotlight. So maybe I'm stating the obvious here. Ah well, better explicit than implicit!)

When I wrote "implementing OROM in the DOM using the JS console" above, I initially forgot about the JS console and had to add it in as part of the implementation technology. I'm sure there is some important difference between the role the DOM plays, and the role the JS console plays, that's relevant to our task.

As always, feel free to pull back from tangents or to say if something should be elsewhere.
